src/main/java/com/hierynomus/smbj/share/NamedPipe.java

```diff
@@ -141,7 +196,15 @@ public int transact(byte[] inBuffer, byte[] outBuffer) {
      * @return the number of bytes written to <code>outBuffer</code>
      */
     public int transact(byte[] inBuffer, int inOffset, int inLength, byte[] outBuffer, int outOffset, int outLength) {
-        return ioctl(FSCTL_PIPE_TRANSCEIVE, true, inBuffer, inOffset, inLength, outBuffer, outOffset, outLength);
+        if (outBuffer.length - outOffset < outLength) {
+            throw new IllegalArgumentException();
+        }
+        final byte[] transactOutBuffer = transact(inBuffer);
+        if (transactOutBuffer.length > outLength) {
+            throw new BufferOverflowException();
```

Comment:
Returning the stream also fits without the need to create a new class.  Do you like the example above (you didn't comment on it)?  Also the proposal was two methods: easy-mode, does it all for you, no risk to the caller, you get everything in one nice byte buffer, and hard-mode, caller gets a stream (or other class such as the proposed Transaction), callers job to read, callers job to close; closing will flush the buffer (didn't put that in the example above, but it can of course be added).